File: frontend/src/services/contract.ts
import apiClient from './api'

// شکل پاسخ قرارداد از سمت بک‌اند (snake_case)
export interface ContractData {
  id: number
  venue_id: number
  start_date: string
  end_date: string
  day_of_week: number
  start_time: string
  recurrence: 'weekly' | 'biweekly' | 'monthly'
  original_price: number
  discounted_price: number
  total_amount: number
  status: string
  description: string | null
}

export interface ContractCreateData {
  venueId: number
  startDate: string
  endDate: string
  dayOfWeek: number
  startTime: string
  recurrence: 'weekly' | 'biweekly' | 'monthly'
  discountedPrice: number
  description?: string
}

export const contractService = {
  // لیست قراردادهای کاربر فعلی
  getAll: async (): Promise<ContractData[]> => {
    const response = await apiClient.get('/contracts')
    return response.data
  },

  // جزئیات یک قرارداد
  getById: async (id: number): Promise<ContractData> => {
    const response = await apiClient.get(`/contracts/${id}`)
    return response.data
  },

  // ثبت قرارداد بلندمدت جدید
  create: async (data: ContractCreateData): Promise<ContractData> => {
    const response = await apiClient.post('/contracts', {
      venue_id: data.venueId,
      start_date: data.startDate,
      end_date: data.endDate,
      day_of_week: data.dayOfWeek,
      start_time: data.startTime,
      recurrence: data.recurrence,
      discounted_price: data.discountedPrice,
      description: data.description || null,
    })
    return response.data
  },
}

File: backend/app/models/contract.py
from sqlmodel import SQLModel, Field, Relationship
from typing import Optional, List
from datetime import date, time, datetime
from enum import Enum

class ContractStatus(str, Enum):
    ACTIVE = "active"
    EXPIRED = "expired"
    CANCELLED = "cancelled"
    SUSPENDED = "suspended"

class PaymentStatus(str, Enum):
    PENDING = "pending"
    PAID = "paid"
    PARTIAL = "partial"
    OVERDUE = "overdue"

class RecurrenceType(str, Enum):
    WEEKLY = "weekly"
    BIWEEKLY = "biweekly"
    MONTHLY = "monthly"

class Contract(SQLModel, table=True):
    __tablename__ = "contracts"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id")
    venue_id: int = Field(foreign_key="venues.id")
    
    start_date: date
    end_date: date
    recurrence: RecurrenceType = Field(default=RecurrenceType.WEEKLY)
    
    day_of_week: int
    start_time: time
    duration: int = Field(default=90)
    
    original_price: int
    discounted_price: int
    total_amount: int
    
    status: ContractStatus = Field(default=ContractStatus.ACTIVE)
    payment_status: PaymentStatus = Field(default=PaymentStatus.PENDING)
    
    description: Optional[str] = None
    auto_renew: bool = Field(default=False)
    
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    user: "User" = Relationship(back_populates="contracts")
    venue: "Venue" = Relationship(back_populates="contracts")
    generated_slots: List["ContractSlot"] = Relationship(back_populates="contract")
    payments: List["ContractPayment"] = Relationship(back_populates="contract")

class ContractSlot(SQLModel, table=True):
    __tablename__ = "contract_slots"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    contract_id: int = Field(foreign_key="contracts.id")
    slot_id: int = Field(foreign_key="slots.id")
    
    session_date: date
    is_attended: bool = Field(default=False)
    is_cancelled: bool = Field(default=False)
    cancellation_reason: Optional[str] = None
    
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    contract: "Contract" = Relationship(back_populates="generated_slots")
    slot: "Slot" = Relationship(back_populates="contract_reference")

class ContractPayment(SQLModel, table=True):
    __tablename__ = "contract_payments"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    contract_id: int = Field(foreign_key="contracts.id")
    
    amount: int
    due_date: date
    paid_at: Optional[datetime] = None
    is_paid: bool = Field(default=False)
    transaction_id: Optional[str] = None
    
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    contract: "Contract" = Relationship(back_populates="payments")
